pahfit/model.py

```diff
@@ -0,0 +1,581 @@
+from specutils import Spectrum1D
+from astropy import units as u
+import copy
+from astropy.modeling.fitting import LevMarLSQFitter
+from matplotlib import pyplot as plt
+import numpy as np
+
+from pahfit.helpers import find_packfile
+from pahfit.features import Features
+from pahfit.base import PAHFITBase
+from pahfit import instrument
+from pahfit.errors import PAHFITModelError
+
+
+class Model:
+    """This class acts as the main API for PAHFIT.
+
+    The users deal with model objects, of which the state is modified
+    during initalization, initial guessing, and fitting. What the model
+    STORES is a description of the physics: what features are there and
+    what are their properties, regardless of the instrument with which
+    those features are observed. The methods provided by this class,
+    form the connection between those physics, and what is observed.
+    During fitting and plotting, those physics are converted into a
+    model for the observation, by applying instrumental parameters from
+    the instrument.py module.
+
+    The main thing that defines a model, is the features table, loaded
+    from a YAML file given to the constructor. After construction, the
+    Model can be edited by accessing the stored features table directly.
+    Changing numbers in this table, is allowed, and the updated numbers
+    will be reflected when the next fit or initial guess happens. At the
+    end of these actions, the fit or guess results are stored in the
+    same table.
+
+    The model can be saved.
+
+    The model can be copied.
+
+    Attributes
+    ----------
+    features : Features
+        Instance of the Features class. Can be edited on-the-fly.
+        Non-breaking behavior by the user is expected. Changes will be
+        reflected at the next fit, guess, or plot call.
+
+    """
+
+    def __init__(self, features: Features):
+        """
+        Parameters
+        ----------
+        features: Features
+            Features table.
+
+        """
+        if len(features) < 2:
+            raise PAHFITModelError(
+                "Fewer than 2 features! Single component models are no allowed!"
+            )
+
+        self.features = features
+        # if set to False, use set fwhm for lines to value in features
+        # table at model construction
+        self.use_instrument_fwhm = True
+
+        # store fit_info dict of last fit
+        self.fit_info = None
+
+    @classmethod
+    def from_yaml(cls, pack_file):
```

Comment:
Oh, I didn't see that `Features.read()` did this too (the docstring does not mention it at the moment). Now that I know that logic was already in place, I am pro
1. Either merging the two "from_*" functions into a single thing called "read"
2. Or calling them "read_pack(yaml_file, instrument)" and "read(_saved?)_model(table_file)" for consistency.

While Features is actually what's being read, we do want to keep a convenience wrapper like this, otherwise the user needs two lines (Features.read + Model constructor), and needs to import more things.)

If we go back to setting the instrument at construction (as we discussed in the past), it might make sense to keep the separate logic (science packs don't have instrument information, but saved models do). So "read_pack" would need an "instrument" argument

There's also the option to move all of this logic to the constructor. I'm mildly against passing file names to the constructor, since I was following the general trend of Astropy-related classes (Table, Spectrum1D, CCDData). They typically have a constructor that takes the data, and a class method called `read()` that takes a file name and calls the constructor for you.

In any case, I will at least remove the duplicate effort here, and delegate the pack-file-finding to Features.

Reply:
Yeah, a Model is not a science pack.  Even if it can pose as one on disk.  I wonder if that will be confusing for people.   I'd let `Features` handle the science pack yaml files, searching for them, etc.  And Model can be responsible for reading Model files only (which yes, are just serialized Features tables).  Maybe on _Model_ save we should add a suffix like `.pfm.yaml`, i.e. for PAHFIT-model, to discriminate.  Or I guess it's `.pfm.ecsv`.  Maybe that's enough of a distinction (since YAML is a "read-only" format here).  What I think is the simplest:

1. When creating a model, you specify the desired science pack.  It's read in with Features.read.  You can also optionally specify instrument(s).
2. When guessing/fitting a model, you can set/update the instrument(s) from the Spectrum1D object (or via keyword).
3. When saving a Model, all the features rows and the instrument details, redshift, etc. get saved (how is not for the user to worry about). 
4. When reading a Model from disk, all that gets loaded in just as it was when you saved it.  Again, how this happens is not a user concern. 

How's that sound?